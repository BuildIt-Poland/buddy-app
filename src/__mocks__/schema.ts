export enum UserRole {
  Newbie = 'NEWBIE',
  Buddy = 'BUDDY',
}

export enum TaskStatus {
  Completed = 'COMPLETED',
  Uncompleted = 'UNCOMPLETED',
}

export default {
  UserRole,
  TaskStatus,
};
