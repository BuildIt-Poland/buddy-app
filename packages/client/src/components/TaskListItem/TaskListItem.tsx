import React from 'react';
import { useLocation, useHistory } from 'react-router-dom';
import { makeStyles } from '@material-ui/core/styles';
import { TaskStatus } from '@buddy-app/schema';
import ListItem from '@material-ui/core/ListItem';
import ListItemIcon from '@material-ui/core/ListItemIcon';
import ListItemText from '@material-ui/core/ListItemText';
import TaskOptions from 'components/TaskOptions';
import TaskCheckbox from 'atoms/TaskCheckbox';
import { TaskListItemProps } from './types';

const useStyles = makeStyles(theme => ({
  listItem: {
    [theme.breakpoints.up('sm')]: {
      paddingLeft: theme.spacing(0.5),
    },
  },
  listOption: {
    justifyContent: 'flex-end',
  },
}));

const TaskListItem: React.FC<TaskListItemProps> = ({
  task,
  tabIndex = 0,
  onChange,
}) => {
  const { id, title, status } = task;
  const { pathname } = useLocation();
  const history = useHistory();
  const { listItem, listOption } = useStyles();
  const text = {
    [TaskStatus.Completed]: <del>{title}</del>,
    [TaskStatus.Uncompleted]: <strong>{title}</strong>,
  };
  const stopPropagation = (e: React.MouseEvent) => e.stopPropagation();
  const listItemClickHandler = () =>
    history.push({ pathname: `${pathname}/${id}`, state: { tabIndex } });

  return (
    <ListItem
      disableGutters
      button
      className={listItem}
      onClick={listItemClickHandler}
      component='li'>
      <ListItemIcon onClick={stopPropagation}>
        <TaskCheckbox task={task} onChange={onChange} />
      </ListItemIcon>
      <ListItemText primary={text[status]} />
      <ListItemIcon className={listOption} onClick={stopPropagation}>
        <TaskOptions id={id} />
      </ListItemIcon>
    </ListItem>
  );
};

export default TaskListItem;
