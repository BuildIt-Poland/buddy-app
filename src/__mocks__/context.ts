import { AuthContextData } from 'context/AuthStore';

export const authContext = (context?: Partial<AuthContextData>) => ({
  data: { role: 'BUDDY', token: 'dummy-token', userId: '1' },
  loading: false,
  error: null,
  isAuthenticated: true,
  login: jest.fn(),
  logout: jest.fn(),
  ...context,
});
