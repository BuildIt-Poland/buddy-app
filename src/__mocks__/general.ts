import { ROUTES } from 'shared/routes';

export const mockLocation = (path: string = ROUTES.BASE) => ({
  key: 'utwyk7',
  pathname: path,
});

export const mockSchemaTypes = (types: any = {}) => ({
  UserRole: {
    Newbie: 'NEWBIE',
    Buddy: 'BUDDY',
  },
  TaskStatus: {
    Completed: 'COMPLETED',
    Uncompleted: 'UNCOMPLETED',
  },
  ...types,
});
