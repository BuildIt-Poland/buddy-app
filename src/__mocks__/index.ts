export * from './general';
export * from './graphql';
export * from './schema';
