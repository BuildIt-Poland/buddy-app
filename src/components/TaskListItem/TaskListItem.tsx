import React from 'react';
import { useLocation, useHistory } from 'react-router-dom';
import { makeStyles } from '@material-ui/core/styles';
import { TaskStatus } from 'buddy-app-schema';
import ListItem from '@material-ui/core/ListItem';
import ListItemIcon from '@material-ui/core/ListItemIcon';
import ListItemText from '@material-ui/core/ListItemText';
import TaskOptions from 'components/TaskOptions';
import TaskCheckbox from 'components/TaskCheckbox';
import { TaskListItemProps } from './types';

const useStyles = makeStyles(theme => ({
  listItem: {
    paddingLeft: theme.spacing(2),
  },
}));

const TaskListItem: React.FC<TaskListItemProps> = ({
  task,
  tabIndex = 0,
  onChange,
}) => {
  const { id, title, status } = task;
  const { pathname } = useLocation();
  const history = useHistory();
  const { listItem } = useStyles();
  const text = {
    [TaskStatus.Completed]: <del>{title}</del>,
    [TaskStatus.Uncompleted]: <strong>{title}</strong>,
  };
  const stopPropagation = (e: React.MouseEvent) => e.stopPropagation();
  const listItemClickHandler = () =>
    history.push({ pathname: `${pathname}/${id}`, state: { tabIndex } });

  return (
    <ListItem
      disableGutters
      button
      className={listItem}
      onClick={listItemClickHandler}
      component='li'>
      <ListItemIcon onClick={stopPropagation}>
        <TaskCheckbox task={task} onChange={onChange} />
      </ListItemIcon>
      <ListItemText primary={text[status]} />
      <ListItemIcon onClick={stopPropagation}>
        <TaskOptions id={id} />
      </ListItemIcon>
    </ListItem>
  );
};

export default TaskListItem;
