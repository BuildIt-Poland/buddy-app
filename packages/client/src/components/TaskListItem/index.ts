import TaskListItem from './TaskListItem';
export * from './types';

export default TaskListItem;
