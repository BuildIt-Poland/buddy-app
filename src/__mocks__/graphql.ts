import { LOGIN_MUTATION } from 'graphql/login.graphql';
import { BUDDY_MENU_DETAILS, NEWBIE_MENU_DETAILS } from 'graphql/user-menu.graphql';
import { AVATAR_HEADER } from 'graphql/avatar-header.graphql';
import { NEWBIE_SELECT } from 'graphql/newbie-select.graphql';
import { TASK_DETAILS } from 'graphql/task-details.graphql';
import { GraphQLError } from 'graphql';
import {
  MutationLoginArgs,
  AuthPayload,
  TaskStatus,
  Newbie,
  QueryNewbieArgs,
  QueryBuddyArgs,
  UserRole,
} from 'buddy-app-schema';
import {
  NEWBIE_CONTACT_DETAILS,
  BUDDY_CONTACT_DETAILS,
} from 'graphql/contact-details.graphql';
import { getBasicUserDetailsMock } from './general';
import { authContext } from './context';

export const loginSuccessMock = (
  variables?: Partial<MutationLoginArgs>,
  data?: Partial<AuthPayload>
) => ({
  request: {
    query: LOGIN_MUTATION,
    variables: {
      email: 'aa@aa.pt',
      password: '12345',
      ...variables,
    },
  },
  result: {
    data: {
      login: {
        ...authContext().data,
        ...data,
      },
    },
  },
});

export const loginFailedMock = (variables?: Partial<MutationLoginArgs>) => ({
  request: {
    query: LOGIN_MUTATION,
    variables: {
      email: 'aa@aa.pt',
      password: '12345',
      ...variables,
    },
  },
  result: {
    errors: [new GraphQLError('No such user found')],
  },
});

export const loginNoNetworkMock = (variables?: Partial<MutationLoginArgs>) => ({
  request: {
    query: LOGIN_MUTATION,
    variables: {
      email: 'aa@aa.pt',
      password: '12345',
      ...variables,
    },
  },
  error: new Error('Network error'),
});

export const NewbieAvatarDetails = (variables?: QueryNewbieArgs) => ({
  request: {
    query: AVATAR_HEADER,
    variables,
  },
  result: {
    data: {
      newbie: {
        name: 'Tom Hanks',
        position: 'front-end',
        photo: 'some-url',
        tasksInfo: {
          buddyProgress: 0.54,
        },
      },
    },
  },
});

export const tasksResponse = (filter?: TaskStatus) => {
  const tasks = [
    {
      id: '1',
      title: 'Call you head manager',
      description: '',
      createdAt: '',
      implementationDate: '',
      status: TaskStatus.Uncompleted,
      newbie: {} as Newbie,
    },
    {
      id: '2',
      title: 'Write you main manager',
      description: '',
      createdAt: '',
      implementationDate: '',
      status: TaskStatus.Completed,
      newbie: {} as Newbie,
    },
  ];

  switch (filter) {
    case TaskStatus.Completed:
      return [tasks[0]];
    case TaskStatus.Uncompleted:
      return [tasks[1]];
    default:
      return tasks;
  }
};

export const buddyMenuDetails = (variables?: Partial<QueryBuddyArgs>) => [
  {
    request: {
      query: BUDDY_MENU_DETAILS,
      variables,
    },
    result: {
      data: {
        buddy: {
          name: 'Tom Hanks',
          email: 'tom@wipro.com',
          allowPushedNotifications: true,
          photo: 'buddy-photo-url',
          newbies: [
            {
              id: 'id',
              name: 'Sandra Bullock',
              photo: 'newbie-photo-url',
            },
          ],
        },
      },
    },
  },
];

export const newbieMenuDetails = (variables?: Partial<QueryNewbieArgs>) => [
  {
    request: {
      query: NEWBIE_MENU_DETAILS,
      variables,
    },
    result: {
      data: {
        newbie: {
          name: 'Sandra Bullock',
          email: 'doejoe@wipro.com',
          allowPushedNotifications: true,
          photo: 'newbie-photo-url',
          buddy: {
            id: 'id',
            name: 'Tom Hanks',
            photo: 'newbie-photo-url',
          },
        },
      },
    },
  },
];

export const newbieContactDetails = (variables?: Partial<QueryNewbieArgs>) => [
  {
    request: {
      query: NEWBIE_CONTACT_DETAILS,
      variables,
    },
    result: {
      data: {
        newbie: getBasicUserDetailsMock(UserRole.Newbie),
      },
    },
  },
];

export const buddyContactDetails = (variables?: Partial<QueryBuddyArgs>) => [
  {
    request: {
      query: BUDDY_CONTACT_DETAILS,
      variables,
    },
    result: {
      data: {
        buddy: getBasicUserDetailsMock(UserRole.Buddy),
      },
    },
  },
];

export const newbieSelectMock = [
  {
    request: {
      query: NEWBIE_SELECT,
      variables: {
        buddyId: 'ck17sl83c9gya0b17dcvttzm4',
      },
    },
    result: {
      data: {
        buddy: {
          id: 'ck17sl83c9gya0b17dcvttzm4',
          name: 'Dummy',
          role: 'BUDDY',
          photo: null,
          newbies: [
            {
              id: 'ck17svulh9k2k0b17j31ansfk',
              photo: null,
              name: 'Test 2',
              startDate: null,
              tasksInfo: { buddyProgress: 0.5 },
            },
            {
              id: 'ck17swp2m9kcv0b17we0ibrdn',
              photo: null,
              name: 'Test 1',
              startDate: null,
              tasksInfo: { buddyProgress: 0.6 },
            },
          ],
        },
      },
    },
  },
];

export const taskDetailsMock = [
  {
    request: {
      query: TASK_DETAILS,
      variables: {
        taskId: '1',
      },
    },
    result: {
      data: {
        task: {
          title: 'New task',
          description: 'New task',
          status: TaskStatus.Uncompleted,
        },
      },
    },
  },
];
