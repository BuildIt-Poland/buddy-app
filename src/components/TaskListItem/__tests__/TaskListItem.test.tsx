import React from 'react';
import { create } from 'react-test-renderer';
import { TaskStatus } from '__mocks__';
import { MemoryRouter } from 'react-router';
import TaskListItem from '../TaskListItem';

jest.mock('@material-ui/core/ListItem', () => 'ListItem');
jest.mock('@material-ui/core/ListItemIcon', () => 'ListItemIcon');
jest.mock('@material-ui/core/ListItemText', () => 'ListItemText');
jest.mock('components/TaskCheckbox', () => 'TaskCheckbox');
jest.mock('components/TaskOptions', () => 'TaskOptions');

describe('Component - TaskCheckbox', () => {
  const { Uncompleted, Completed } = TaskStatus;

  test(`renders correctly when task is ${Uncompleted}`, () => {
    const component = create(
      <MemoryRouter>
        <TaskListItem
          title={'Change you line manager'}
          status={Uncompleted}
          id={'1'}
          onChange={() => null}
        />
      </MemoryRouter>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
  test(`renders correctly when task is ${Completed}`, () => {
    const component = create(
      <MemoryRouter>
        <TaskListItem
          title={'Change you line manager'}
          status={Completed}
          id={'1'}
          onChange={() => null}
        />
      </MemoryRouter>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});
