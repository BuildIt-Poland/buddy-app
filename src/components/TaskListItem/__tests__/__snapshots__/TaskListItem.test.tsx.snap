// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component - TaskCheckbox renders correctly when task is COMPLETED 1`] = `
<ListItem
  button={true}
  className="makeStyles-listItem-1"
  component="li"
  disableGutters={true}
  onClick={[Function]}
>
  <ListItemIcon
    onClick={[Function]}
  >
    <TaskCheckbox
      id="1"
      onChange={[Function]}
      status="COMPLETED"
    />
  </ListItemIcon>
  <ListItemText
    primary={
      <del>
        Change you line manager
      </del>
    }
  />
  <ListItemIcon
    onClick={[Function]}
  >
    <TaskOptions
      id="1"
    />
  </ListItemIcon>
</ListItem>
`;

exports[`Component - TaskCheckbox renders correctly when task is UNCOMPLETED 1`] = `
<ListItem
  button={true}
  className="makeStyles-listItem-1"
  component="li"
  disableGutters={true}
  onClick={[Function]}
>
  <ListItemIcon
    onClick={[Function]}
  >
    <TaskCheckbox
      id="1"
      onChange={[Function]}
      status="UNCOMPLETED"
    />
  </ListItemIcon>
  <ListItemText
    primary={
      <strong>
        Change you line manager
      </strong>
    }
  />
  <ListItemIcon
    onClick={[Function]}
  >
    <TaskOptions
      id="1"
    />
  </ListItemIcon>
</ListItem>
`;
